spotfinder/kernels/thresholding.cu

```diff
@@ -0,0 +1,323 @@
+                                         size_t result_pitch,
+                                         int width,
+                                         int height,
+                                         pixel_t max_valid_pixel_value,
+                                         uint8_t kernel_width,
+                                         uint8_t kernel_height,
+                                         uint8_t min_count,
+                                         float n_sig_b,
+                                         float n_sig_s) {
+    // Move pointers to the correct slice
+    image = image + (image_pitch * height * blockIdx.z);
+    result_mask = result_mask + (mask_pitch * height * blockIdx.z);
+
+    // Calculate the pixel coordinates
+    auto block = cg::this_thread_block();
+    int x = block.group_index().x * block.group_dim().x + block.thread_index().x;
+    int y = block.group_index().y * block.group_dim().y + block.thread_index().y;
+
+    if (x >= width || y >= height) return;  // Out of bounds guard
+
+    pixel_t this_pixel = image[y * image_pitch + x];
+
+    // Check if the pixel is masked and below the maximum valid pixel value
+    bool px_is_valid =
+      mask[y * mask_pitch + x] != 0 && this_pixel <= max_valid_pixel_value;
+
+    // Initialize variables for computing the local sum and count
+    uint sum = 0;
+    size_t sumsq = 0;
+    uint8_t n = 0;
+
+    int row_start = max(0, y - kernel_height);
+    int row_end = min(y + kernel_height + 1, height);
+
+    for (int row = row_start; row < row_end; ++row) {
+        int row_offset = image_pitch * row;
+        int mask_offset = mask_pitch * row;
+
+        int col_start = max(0, x - kernel_width);
+        int col_end = min(x + kernel_width + 1, width);
+
+        for (int col = col_start; col < col_end; ++col) {
+            pixel_t pixel = image[row_offset + col];
+            uint8_t mask_pixel = mask[mask_offset + col];
+            bool include_pixel = mask_pixel != 0;  // If the pixel is valid
+            if (include_pixel) {
+                sum += pixel;
+                sumsq += pixel * pixel;
+                n += 1;
+            }
+        }
+    }
+
+    if (px_is_valid && n > 1) {
+        // Compute local mean and variance
+        float sum_f = static_cast<float>(sum);
+        float sumsq_f = static_cast<float>(sumsq);
+
+        float mean = sum_f / n;
+        float variance = (n * sumsq_f - (sum_f * sum_f)) / (n * (n - 1));
+        float dispersion = variance / mean;
+
+        // Compute the background threshold and signal threshold
+        float background_threshold = 1 + n_sig_b * sqrt(2.0f / (n - 1));
+        bool not_background = dispersion > background_threshold;
+        float signal_threshold = mean + n_sig_s * sqrt(mean);
+
+        // Check if the pixel is a strong pixel
+        bool is_signal = this_pixel > signal_threshold;
+
+        result_mask[x + result_pitch * y] = not_background && is_signal;
```

Comment:
Is the addition of `&& is signal` the only difference between these, or is it subtly different otherwise? I wonder if there is a way to unify these a bit, although we don't have to do that now.

Reply:
Yes, this is the _only_ difference between these two. Which is why I had mistakenly used regular dispersion as the first step initially.

All three of these thresholding kernels share a lot of redundant code that I want to unify, but here I wrote them out extremely explicitly while trying to ensure that they matched the DIALS results.

I want to/will remove the redundancies here, however, for this PR I'd like to keep them verbose like this while I work on memory optimisations.

Reply:
I have modularised the most egregious redundancies: 13f7623b5a53a7263211163005f771a85fa19297

Reply:
So this works with `cuda::std::tuple`: 02a0ef6dd6e74b7e0be8927730aa841ee1b0ee17

However, thought it compiles and runs, it does give en error… Something I'd like to talk about.